keps/sig-storage/2268-non-graceful-shutdown/README.md

```diff
@@ -0,0 +1,466 @@
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+In case of a node being shutdown (i.e. not running) the control plane doesn't make the right assumptions to enable stateful workload to fail-over. This KEP introduces a flow to do so.
+
+## Motivation
+
+Today kubernetes doesn't enable storage providers that are able to safely detach volumes (i.e. only detach when volumes aren't being written to) to do so. This proposal aims to introduce a new option to safely trigger volume detach if a node is shutdown.
+
+### Use Cases
+
+* User wants a stateful workload to failover to another running node in case a node is shutdown without any intervention when it's safe to do so.
+
+### Goals
+
+* Increase the availability of stateful workloads when a node is shutdown.
+* Automate self-healing for stateful workloads.
+* This proposal only target CSI volumes.
+
+### Non-Goals
+
+* Node/control plane partitioning other than a node shutdown is not covered by the proposal, but will be addressed in the future and built on top of this design.
+* Implement in-cluster logic to handle node/control plane partitioning. 
+* Enable detach for all the storage providers.
+* Existing in-tree volumes are not targeted by this proposal.
+
+## Proposal
+
+This proposal introduces a new option in CSIDriver. If set to true, it would trigger volume detach if a node is shutdown.
+
+* In the [CSIDriverSpec](https://github.com/kubernetes/kubernetes/blob/v1.20.2/pkg/apis/storage/types.go#L266), add a `SafeDetach` boolean to indicate if it is safe to detach.
+
+```go
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+  ...
+
+  // SafeDetach indicates this CSI volume driver is able to perform detach
+  // operations. When set to true, the CSI volume driver needs to ensure that 
+  // ControllerUnpublishVolume() implementation is checking whether a volume
+  // can be detached safely
+  // +optional
+  SafeDetach *bool
+}
+```
+
+* The Graceful Node Shutdown [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2000-graceful-node-shutdown) introduced a way to detect a node shutdown. When the NodeShutdownManager detects a node is shutting down, it sets the `Reason` field of a `PodStatus` to a constant named `nodeShutdownReason`. Currently `nodeShutdownReason` is in the `nodeshutdown` package. We need to export this constant by renaming it to `NodeShutdownReason` and move it to a package that can be imported by the Pod GC Controller.
```

Comment:
I think there is some confusion to Graceful Node Shutdown. Graceful Node Shutdown largely thinks the node is going down for a reboot and will come back; not be shutdown permanently. If a node is upgrading, then this proposal would detach all the volumes, then cause the Kubelet to thundering-heard attach the volumes once the node comes back up.

Another issue will be network partitions. The Kubelet and Control Plane are designed to be resilient to network partitions. This proposal side steps the network partition question, which is a valid consideration since all pods are assumed to still be running on a Partitioned node. This Partitioned Node could still be writing to a volume.

Regarding
> After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.

The pods in the API may be stuck in terminating state, but the pods could still be running. Once the network partition is resolved, then the API statuses will be updated correctly. The scheduler will then make a determination if the pods on the resolved node should be terminated.

Reply:
> Graceful Node Shutdown largely thinks the node is going down for a reboot and will come back; not be shutdown permanently.

I'm not sure if this is correct assumption. I think with graceful node shutdown we don't really know if the node is going to reboot or will shutdown permanently. All graceful node shutdown knows is that there is a system shutdown imminent, and it delays it to terminate pods gracefully. 

/cc @mrunalp

Reply:
True. After thinking about it overnight, that point of mine does not make sense, but the others are valid.